File: cheapest-flights-within-k-stops/cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        dist = [[1000000000]*102 for _ in range(102)]
        adj = collections.defaultdict(list)
        for f in flights:
            adj[f[0]].append((f[1], f[2]))
        q = []
        heapq.heappush(q, (0, src, 0))
        dist[src][0] = 0
        while q:
            cost, now, cnt = heapq.heappop(q)
            if cnt > k:
                continue
            for next, c in adj[now]:
                if dist[next][cnt + 1] > c + cost:
                    dist[next][cnt + 1] = c + cost
                    heapq.heappush(q, (c + cost, next, cnt + 1))
        ans = dist[dst][0]
        for i in range(1, k + 2):
            ans = ans if dist[dst][i] > ans else dist[dst][i]
        if ans == 1000000000:
            return -1
        return ans
            

File: cheapest-flights-within-k-stops/cheapest-flights-within-k-stops.cpp
class Solution {
public:
    struct D {
        int now, cost, cnt;
        bool operator<(const D& a) const {
            return cost > a.cost;
        }
    };
    vector<pair<int, int>>adj[101];
    long long d[101][103];
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        for(auto& v:flights) {
            int from = v[0];
            int to = v[1];
            int c = v[2];
            adj[from].emplace_back(to, c);
            //adj[to].emplace_back(from, c);
        }
        for(int i=0;i<100;++i){
            for(int j=0;j<101;++j){
                d[i][j] = 1000000000;
            }
        }
        priority_queue<D>q;
        q.push({src, 0, 0});
        d[src][0] = 0;
        while(!q.empty()) {
            int now = q.top().now;
            int c = q.top().cost;
            int cnt = q.top().cnt;
            q.pop();
            if(d[now][cnt] < c) continue;
            for(auto p : adj[now]) {
                int next = p.first;
                int nc = p.second + c;
                if(d[next][cnt + 1] > nc) {
                    d[next][cnt+1] = nc;
                    q.push({next, nc, cnt+1});
                }
            }
        }
        int ans = 1000000000;
        for(int i=0;i<=k+1;++i) {
            if(d[dst][i] < ans) {
                ans = d[dst][i];
            }
        }
        if(ans==1000000000) ans = -1;
        return ans;
        
    }
};